OpenZFS 2.1.1 patch set

behlendorf opened the pull request:
### Motivation and Context

Proposed patch set for the OpenZFS 2.1.1 point release.  PR opened for CI testing and code review.  All patches were cherry-picked from the master branch and apply cleanly.

### Description

#### Bug Fixes

- c015fcc25 Fix/improve dbuf hits accounting
- 337b92d36 Restore FreeBSD sysctl processing for arc.min and arc.max
- 53416da56 Enable /proc/diskstats for zvols
- 204507d36 Read past end of argv array in zpool_do_import()
- 22339052c Fixes in persistent L2ARC
- f7b2b2de7 Fix unfortunate NULL in spa_update_dspace
- e4dada50f dmu_redact.c does not call bqueue_destroy
- d06bd2fa3 config/libatomic: require -latomic iff atomic.c doesn't link w/o it
- 18ce2f2d7 Tinker with slop space accounting with dedup
- 536cf83b5 Fix ARC ghost states eviction accounting
- e53a97a5b file reference counts can get corrupted
- 1d2b2ac0c Use substantially more robust program exit status logic in zvol_id
- a76101b5d Udev rules: use match (==) rather than assign (=) for PROGRAM
- 05ad10a43 Udev rules: replace deprecated $tempnode with $devnode
- 8c6a571f5 Udev rules: use non-ancient comma syntax
- 841e76e32 Fix abd leak, kmem_free correct size of abd_t
- cb1a5ad8b Upstream: dmu_zfetch_stream_fini leaks refcount
- 724a1f442 ZED: Match added disk by pool/vdev GUID if found
- bae37f15c Modify checksum obtain method of QAT

#### Build Fixes (Linux/FreeBSD kernel changes, compiler warnings, build system, etc)

- ab88553f3 Initialize parity blocks before RAID-Z reconstruction benchmarking
- f194cfd42 Linux 4.11 compat: statx support
- ad773bb3e zfs: add missed dependency of zfs module on zlib
- de18c9121 Fix usage of find in tests/Makefile.am
- 7bc11c03b Initialize dn_next_type[] in the dnode constructor
- ce6cf8a86 Zero pad bytes following TX_WRITE log data
- 18adb20fe Zero pad bytes when allocating a ZIL record
- 1ad1fe4a7 Initialize all fields in zfs_log_xvattr()
- 0295c7033 Initialize "autoreplace" in spa_ld_get_props()
- 776fd9b5a Linux 5.14 compat: explicity assign set_page_dirty
- eb17f92e1 Linux 5.14 compat: blk_alloc_disk()
- 5c0f66bf6 FreeBSD: Ignore make_dev_s() errors
- f98ec65f4 Remove old orig_fd variable from zfs send
- f8d265935 CI: generate ABI files if changed
- 1cf3e3c27 A few fixes of callback typecasting (for the upcoming ClangCFI)
- ba7b76892 Remove unused fields from zvol_task_t
- 087504936 FreeBSD: Switch from MAXPHYS to maxphys on FreeBSD 13+
- 4f79b0429 zpool_influxdb: fix -Werror=stringop-truncation
- 41bef368f Update bug report template
- a3e12cbc6 dprintf_dnode: strcpy -> strlcpy
- fe4efe6bc Replace strchrnul() with strrchr()
- ec6a6e869 Detect HAVE_LARGE_STACKS at compile time

#### Performance

- 1cc07ce45 Remove b_pabd/b_rabd allocation from arc_hdr_alloc()
- ad944f4f8 Optimize arc_l2c_only lists assertions
- 8c207f116 Avoid vq_lock drop in vdev_queue_aggregate()
- 10d7bfbf6 Use more atomics in refcounts
- 8c7d7121d Run arc_evict thread at higher priority
- 21ae63695 Allow disabling of unmapped I/O on FreeBSD
- feaa7f503 Avoid small buffer copying on write
- 9a2b63a78 Optimize allocation throttling
- d790c5c83 Minor ARC optimizations
- 45b0192c4 Introduce dsl_dir_diduse_transfer_space()
- 00d037b5b FreeBSD: Use unmapped I/O for scattered/gang ABD buffers
- 905b663b6 FreeBSD: Hardcode abd_chunk_size to PAGE_SIZE
- afa0a2934 Move gethrtime() calls out of vdev queue lock
- d339ca52d Compact dbuf/buf hashes and lock arrays
- cf5b67ef5 Remove refcount from spa_config_*()
- cae1bfc09 Optimize small random numbers generation

#### Documentation

- aff1d4528 zpool-remove.8: describe top-level vdev sector size limitation
- 0beddaac4 zfs.4: Fix typo s/compatiblity/compatibility/
- b7aa09c54 Make get_key_material_file fail more verbosely
- 85cabd996 Man zpool-scrub.8: describe sequential scrub
- 8c8064197 Add missing properties to zfs allow manpage
- e12ed15d2 Add comment on metaslab_class_throttle_reserve() locking
- 9d6ca2665 Correct zfs-send(8) on readonly sends
- 956ad97cd Print zvol_id error messages to stderr rather than stdout

#### ZFS Test Suite

- e1866266b ZTS: Add tests for creation time
- 7ea620c65 ZTS: Avoid unset $tmpdir in redacted_panic
- 733a8fee4 Add zfs.sh -r flag to reload modules
- 696af5e23 zloop: Add a max iterations option, use default run/pass times
- 9360ea1da Assorted fixes for the performance tests

### How Has This Been Tested?

Local run of the ZTS using the sanity runfile on Ubuntu 18.04.  Additional testing to be performed by the CI for supported platforms.

```sh
Results Summary
PASS     811

Running Time:   00:17:08
Percent passed: 100.0%
Log directory:  /var/tmp/test_results/20210823T223519

Tests with results other than PASS that are expected:

Tests with result of PASS that are unexpected:

Tests with results other than PASS that are unexpected:
```

### Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Performance enhancement (non-breaking change which improves efficiency)
- [x] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Library ABI change (libzfs, libzfs\_core, libnvpair, libuutil and libzfsbootenv)
- [x] Documentation (a change to man pages or other documentation)

### Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the OpenZFS [code style requirements](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md#coding-conventions).
- [x] I have updated the documentation accordingly.
- [x] I have read the [**contributing** document](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md).
- [x] I have added [tests](https://github.com/openzfs/zfs/tree/master/tests) to cover my changes.
- [x] I have run the ZFS Test Suite with this change applied.
- [x] All commit messages are properly formatted and contain [`Signed-off-by`](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md#signed-off-by).

behlendorf commented:
@freqlabs would you mind looking this over to make sure we have everything needed for FreeBSD.  I did skip a couple FreeBSD related changes were performed some code refactoring or other slightly invasive optimization.

freqlabs commented:
Looks good on the FreeBSD side. I appreciate that you pulled in the unmapped IO changes for FreeBSD, as well.

I do see a few more simple general fixes and optimizations that would be nice to have:

#### Bug Fixes

- ~~cfe8e960f193d69a87693474de666025736f16aa Fix/improve dbuf hits accounting~~
- e829a865bf6482a0d62913f528e516ce3b759d7b Use more atomics in refcounts

#### Performance Improvements

- ~~6b88b4b501a9e12dd6feee7e3cb2824bc70beeca Remove b_pabd/b_rabd allocation from arc_hdr_alloc()~~
- ~~bb7ad5d326a0b3d7c327e9edd1bc4fcb9b72ff82 Optimize arc_l2c_only lists assertions~~
- 7f9d9e6f39fdcbd048b5f9991186da2e14d8139c Avoid vq_lock drop in vdev_queue_aggregate()
- 1b50749ce9757030ddd4dba9f03ee18182cda82e Optimize allocation throttling
- ~~42afb12da70f1fd73c7fcf653d5e2bf42f05b42b Remove refcount from spa_config_\*()~~

You got a few of these while I was still writing. :smile:

behlendorf commented:
@freqlabs thanks for the additional suggestions.  I think it's reasonable to pull in those changes so I've added them to this PR (and updated the top commit).  I also pulled in a couple of other small but nice to have performance improvements.
